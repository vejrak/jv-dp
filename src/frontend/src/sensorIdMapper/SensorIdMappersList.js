// @flow

import React from 'react'
import { Button } from 'react-bootstrap'
import filter from 'lodash/filter'
import {
  deleteSensorIdMapper,
  editSensorIdMapper,
  getSensorIdMappers,
} from './actions'
import { canShowUpdateForm } from '../helpers'
import type { ErrorResponse, SensorIdMapper } from '../types'
import SensorIdMapperForm from './form/SensorIdMapperFormConnected'
import Spinner from '../components/Spinner'

type Props = $ReadOnly<{|
  deleteSensorIdMapper: typeof deleteSensorIdMapper,
  editSensorIdMapper: typeof editSensorIdMapper,
  getSensorIdMappers: typeof getSensorIdMappers,
  isFetching: boolean,
  error: ?ErrorResponse,
  sensorIdMappers: Array<SensorIdMapper>,
|}>

type State = {
  showEditFormsFor: Array<string>,
}

class SensorIdMappersList extends React.PureComponent<Props, State> {
  state = {
    showEditFormsFor: [],
  }

  componentDidMount() {
    this.props.getSensorIdMappers()
  }

  handleEditClick = (_id: string): void => {
    const { showEditFormsFor } = this.state
    this.setState({ showEditFormsFor: showEditFormsFor.concat(_id) })
  }

  handleHideClick = (_id: string): void => {
    const { showEditFormsFor } = this.state
    this.setState({
      showEditFormsFor: filter(showEditFormsFor, _id),
    })
  }

  handleDeleteClick = async (_id: string) => {
    await this.props.deleteSensorIdMapper(_id)
  }

  updateSensorIdMapper = (_id: string): Function => (
    sensorIdMapper: SensorIdMapper,
  ): void => {
    this.props.editSensorIdMapper(sensorIdMapper, _id)
  }

  render() {
    const { showEditFormsFor } = this.state
    const { error, isFetching, sensorIdMappers } = this.props

    if (isFetching) return <Spinner />

    if (error)
      return (
        <div className="alert alert-danger" role="alert">
          {error.errmsg}
        </div>
      )

    if (!sensorIdMappers || sensorIdMappers.length === 0) return null
    return (
      <table className="table">
        <thead>
          <tr>
            <th scope="col">Name</th>
            <th scope="col">Content type</th>
            <th />
            <th />
          </tr>
        </thead>
        {sensorIdMappers.map((sensorIdMapper) => (
          <tbody key={sensorIdMapper._id} className="mb-4">
            <tr>
              <td>{sensorIdMapper.source_name}</td>
              <td>{sensorIdMapper.content_type}</td>
              <td>
                <Button
                  className="mt-3 mb-3 btn d-block btn-danger"
                  onClick={() => {
                    if (
                      window.confirm(
                        'Are you sure you wish to delete this item?',
                      )
                    )
                      this.handleDeleteClick(sensorIdMapper._id)
                  }}
                >
                  Delete
                </Button>
              </td>
              {!canShowUpdateForm(showEditFormsFor, sensorIdMapper._id) && (
                <td>
                  <Button
                    className="mt-3 mb-3"
                    onClick={() => this.handleEditClick(sensorIdMapper._id)}
                  >
                    Edit
                  </Button>
                </td>
              )}
            </tr>
            {canShowUpdateForm(showEditFormsFor, sensorIdMapper._id) && (
              <>
                <tr>
                  <td colSpan={2}>
                    <SensorIdMapperForm
                      sensorIdMapper={sensorIdMapper}
                      updateSensorIdMapper={this.updateSensorIdMapper(
                        sensorIdMapper._id,
                      )}
                    />
                  </td>
                  <td>
                    <Button
                      className="mt-3 mb-3"
                      onClick={() => this.handleHideClick(sensorIdMapper._id)}
                    >
                      Hide
                    </Button>
                  </td>
                </tr>
              </>
            )}
          </tbody>
        ))}
      </table>
    )
  }
}

export default SensorIdMappersList
