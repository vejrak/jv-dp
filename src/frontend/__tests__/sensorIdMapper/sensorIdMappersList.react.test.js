import React from 'react'
import renderer from 'react-test-renderer'
import SensorIdMappersList from '../../src/sensorIdMapper/SensorIdMappersList'

const mockProps = {
  deleteSensorIdMapper: () => {},
  editSensorIdMapper: () => {},
  getSensorIdMappers: () => {},
  isFetching: false,
  error: null,
  sensorIdMappers: [
    { _id: 'id1', content_type: 'json', source_name: 'key.value' },
  ],
}

it('renders correctly', () => {
  const tree = renderer.create(<SensorIdMappersList {...mockProps} />).toJSON()
  expect(tree).toMatchSnapshot()
})
