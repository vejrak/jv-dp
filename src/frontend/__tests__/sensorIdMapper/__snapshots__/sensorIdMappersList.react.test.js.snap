// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<table
  className="table"
>
  <thead>
    <tr>
      <th
        scope="col"
      >
        Name
      </th>
      <th
        scope="col"
      >
        Content type
      </th>
      <th />
      <th />
    </tr>
  </thead>
  <tbody
    className="mb-4"
  >
    <tr>
      <td>
        key.value
      </td>
      <td>
        json
      </td>
      <td>
        <button
          className="mt-3 mb-3 btn d-block btn-danger btn btn-primary"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Delete
        </button>
      </td>
      <td>
        <button
          className="mt-3 mb-3 btn btn-primary"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Edit
        </button>
      </td>
    </tr>
  </tbody>
</table>
`;
